main.go

```diff
@@ -0,0 +1,335 @@
+}
+type contact_info struct{
+	Name   		string   `db:"name"`
+	Number 		string   `db:"number"`
+	Email  		string   `db:"email"`
+	Nationality 	string   `db:"nationality"`
+	Address 	string   `db:"address"`
+	Username  	string   `db:"username"`
+	ID 		int   `db:"id"`
+}
+type LoginPage struct {
+	Error string
+}
+type page struct{
+	Contacts []contact_info
+	Numbers []Telephone
+}
+type Telephone struct {
+	Contact_id int `db:"contactid"`
+	Number string  	`db:"number"`
+	Num_id int	`db:"numid"`
+
+}
+var muxer = mux.NewRouter()
+func main(){
+	initDb()
+
+	var contact contact_info
+	var userAccount User
+	var telephone Telephone
+
+	muxer.HandleFunc("/viewNumbers/{id}", telephone.ViewTelephonesHandler).Methods("GET")
+	muxer.HandleFunc("/contact",contact.SaveContactHandler).Methods("POST")
+	muxer.HandleFunc("/contact/{id}" , contact.DeleteContactHandler).Methods("DELETE")
+	muxer.HandleFunc("/deleteNumber/{numId}" , telephone.DeleteNumberHandler).Methods("DELETE")
+	muxer.HandleFunc("/addNumber/{ContactId}" ,telephone.AddNumberHandler).Methods("POST")
+ 	muxer.HandleFunc("/logout" ,userAccount.logoutHandler)
+	muxer.HandleFunc("/" ,serverContent)
+
+	muxer.HandleFunc("/login" ,userAccount.loginContactHandler)
+	muxer.HandleFunc("/{page_alias}" , serverContent)
+
+	muxer.HandleFunc("/img/" ,serverResource)
+	muxer.HandleFunc("/js/" ,serverResource)
+	muxer.HandleFunc("/css/{page_alias}" ,serverResource)
+
+	//it provides some default middleware
+	n := negroni.Classic()
+	n.Use(sessions.Sessions("go-for-web-dev" , cookiestore.New([]byte("my-secret-123"))))
+	//add Handler to middleware stack
+	n.Use(negroni.HandlerFunc(verifyDatabase))
+	// to add http.Handler (process that runs in response to request made to web app.)alli f el mux in negroni stack
+	n.UseHandler(muxer)
+	n.Run(":8080")
+	defer db.Close()
+}
+var db *sql.DB
+//var dbmap *gorp.DbMap
+func initDb (){
+	db,_ =sql.Open("mysql" , "root:shahenda_hassan@/mydatabase")
+      	//first parameter is a pointer to our database , second is used sql (sqlight or mysql)
+	//dbmap = &gorp.DbMap{Db:db , Dialect:gorp.MySQLDialect{"InnoDB", "UTF8"}}
+
+	//dbmap.AddTableWithName(contact_info{} , "Contacts").SetKeys(false , "name")
+	//dbmap.AddTableWithName(Telephone{} , "Telephones").SetKeys(false , "number")
+	//dbmap.AddTableWithName(User{} , "Users").SetKeys(false , "username")
+
+	//dbmap.CreateTablesIfNotExists()
+}
+
+func verifyDatabase(w http.ResponseWriter , r *http.Request , next http.HandlerFunc){
+	//ping() -> it verify connection to database is still alive , establishing connection if necessary
+	err := db.Ping();
+	if err != nil{
+		http.Error(w, err.Error() , http.StatusInternalServerError)
+		return
+	}
+	next(w,r)
+}
+
+func(info User)loginContactHandler (w http.ResponseWriter , r *http.Request){
+	var p LoginPage
+
+	if r.FormValue("signUp") != ""{
+		secret , _ := bcrypt.GenerateFromPassword([]byte(r.FormValue("password")) , bcrypt.DefaultCost)
```

Comment:
as bcrypt is the library that generated a salt-password hash ,as each password is hashed the same way , if two users have the same password , they will have the same hash , we can prevent this by randomizing each hash , by appending a random string called salt ,which is stored in user account database .